File: elixir/darts/lib/darts.ex
defmodule Darts do
  @type position :: {number, number}

  defp distance_between( {ax, ay}, {bx, by} ) do
    dx = ax - bx
    dy = ay - by
    Float.pow( Float.pow( dx, 2 ) + Float.pow( dy, 2 ), 1/2 )
  end

  defp distance_from_center( position ) do
    distance_between( position, { 0.0, 0.0 } )
  end

  defp in_circle?( position, radius ) do
    not( distance_from_center( position ) > radius )
  end

  defp scores?( checked_score, position ) do
    cond do
      checked_score == 10 -> in_circle?( position, 1 )
      checked_score == 5 -> in_circle?( position, 5 )
      checked_score == 1 -> in_circle?( position, 10 )
      checked_score == 0 -> true
    end
  end

  @doc """
  Calculate the score of a single dart hitting a target
  """
  @spec score(position) :: integer
  def score({x, y}) do
    cond do
      scores?( 10, { x, y } ) -> 10
      scores?( 5,  { x, y } ) -> 5
      scores?( 1,  { x, y } ) -> 1
      scores?( 0,  { x, y } ) -> 0
    end
  end
end

File: typescript/darts/darts.ts
export function score(x: number, y: number): number {
  const landedFromCenter = distanceFromCenter(x, y);
  return (
    Object.values(targetCircles).find(({ size }) => landedFromCenter <= size)
      ?.score ?? scoreForOutside
  );
}

function distanceFromCenter(x: number, y: number): number {
  return (x ** 2 + y ** 2) ** (1 / 2);
}

const targetCircles = {
  inner: { size: 1, score: 10 },
  middle: { size: 5, score: 5 },
  outer: { size: 10, score: 1 },
} as const;

const scoreForOutside: number = 0;
